common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1155,6 +1156,71 @@ public UTF8String translate(Map<String, String> dict) {
     return fromString(sb.toString());
   }
 
+  public UTF8String translate(Map<String, String> dict, int collationId) {
+    if(CollationFactory.fetchCollation(collationId).supportsBinaryEquality) {
+      return translate(dict);
+    }
+    return collationAwareTranslate(dict, collationId);
+  }
+
+  public UTF8String collationAwareTranslate(Map<String, String> dict, int collationId) {
+    if (numBytes == 0) {
+      return this;
+    }
+
+    Map<String, String> collationAwareDict = getCollationAwareDict(dict, collationId);
```

Comment:
can we just call regular translate with this new collationAwareDict?

Reply:
I think you are right, let me test this.